### PR TITLE
MRG+1: serialize labels for ICA, plot labels and fix ECG guessing (broken)

### PR DESCRIPTION
**Pull request by dengemann**
cc @Eric89GXL @agramfort @jona-sassenhagen 

work in progress.

### CODE REVIEW

**Comment by dengemann** [comment]
cc @jaeilepp FYI

File: mne/preprocessing/ecg.py [hunk 1]
```diff
@@ -189,7 +196,10 @@ def find_ecg_events(raw, event_id=999, ch_name=None, tstart=0.0,
     ecg_events = np.array([ecg_events + raw.first_samp,
                            np.zeros(n_events, int),
                            event_id * np.ones(n_events, int)]).T
-    return ecg_events, idx_ecg, average_pulse
+    out = (ecg_events, idx_ecg, average_pulse)
+    if return_ecg:
+        out += (ecg if created_ecg else None,)
```
Comment (agramfort) [inline mne/preprocessing/ecg.py]: I would return the real ecg if return_ecg == True.

File: mne/preprocessing/ecg.py [hunk 2]
```diff
@@ -278,18 +288,35 @@ def create_ecg_epochs(raw, ch_name=None, event_id=999, picks=None,
     ecg_epochs : instance of Epochs
         Data epoched around ECG r-peaks.
     """
-
-    events, _, _ = find_ecg_events(raw, ch_name=ch_name, event_id=event_id,
-                                   l_freq=l_freq, h_freq=h_freq,
-                                   verbose=verbose)
+    if 'ecg' not in raw:
+        ecg, times = _make_ecg(raw, None, None, verbose)
```
Comment (agramfort) [inline mne/preprocessing/ecg.py]: insert empty line

File: mne/preprocessing/ecg.py [hunk 3]
```diff
@@ -278,18 +288,35 @@ def create_ecg_epochs(raw, ch_name=None, event_id=999, picks=None,
...
+    events, _, _, ecg = find_ecg_events(
+        raw, ch_name=ch_name, event_id=event_id, l_freq=l_freq, h_freq=h_freq,
+        return_ecg=True,
+        verbose=verbose)
```
Comment (agramfort) [inline mne/preprocessing/ecg.py]: insert empty line

File: mne/viz/ica.py [hunk 4]
```diff
@@ -327,7 +365,22 @@ def plot_ica_scores(ica, scores, exclude=None, axhline=None,
...
+    elif labels == 'eog':
+        labels = [l for l in ica.labels_ if l.startswith('eog/')]
+        labels.sort(key=lambda l: l.split('/')[1])  # sort by index
+    elif isinstance(labels, string_types):
+        if len(axes) > 1:
+            raise ValueError('Need as amny labels as axes (%i)' % len(axes))
+        labels = [labels]
+    elif isinstance(labels, (tuple, list)):
+        if len(labels) != len(axes):
+            raise ValueError('Need as amny labels as axes (%i)' % len(axes))
```
Comment (jona-sassenhagen) [inline mne/viz/ica.py]: *many (also in 376)

File: mne/viz/ica.py [hunk 5]
```diff
@@ -327,7 +365,22 @@ def plot_ica_scores(ica, scores, exclude=None, axhline=None,
     else:
         axes = [axes]
     plt.suptitle(title)
-    for this_scores, ax in zip(scores, axes):
+
+    if labels == 'ecg':
+        labels = [l for l in ica.labels_ if l.startswith('ecg')]
+    elif labels == 'eog':
+        labels = [l for l in ica.labels_ if l.startswith('eog/')]
```
Comment (jona-sassenhagen) [inline mne/viz/ica.py]: Do we need the backslash? I assume the idea is to look for something like "eog/blinks", but for many experiments/layouts (e.g. sparse sensor coverage, short visual experiment), there is only one type of eye component.

Comment (dengemann) [inline mne/viz/ica.py]: it has a very particular meaning here. we use it here to access info about the EOG channel or the ECG channel it's private API.

Comment (dengemann) [inline mne/viz/ica.py]: In a score plot you then get the EOG channel.
In other words, users are not supposed to use '/' in their labels.

Comment (jona-sassenhagen) [inline mne/viz/ica.py]: Ah ok.

**Comment by dengemann** [comment]
Some of the viz possible with these changes can be looked at here:

https://www.dropbox.com/s/j8r2ujbnsgq5njr/preprocessing-report-sample.html?dl=0

generated using https://github.com/dengemann/meeg-preprocessing

File: mne/viz/ica.py (hunk 5 above)
Comment (dengemann) [inline mne/viz/ica.py]: I'm fine using a different separator, e.g. a dollar or semicolon, etc. and protecting it (throwing erorrs) if you think that's a good idea.

Comment (jona-sassenhagen) [inline mne/viz/ica.py]: >protecting it (throwing errors)

I think so, yes.

**Comment by dengemann** [comment]
I should summarize the changes maybe to make it easier for you to follow the changes:

- ica labels are now serialized
- the ica find_ecg/eog methods will create sub-labels + store the order to allow for fine grained info during plotting
- the plot_scores and plot_sources(evoked) methods now take advantage of the labeling
- plot_scores has a new argument called labels which take either labels or codes such as 'eog', the latter will then look up the channels used for detection and write them in the figure
- plot_sources(evoked) will look up labels and show them next to the component in the line plot, color goes label-wise
- the ecg API is imprvoed and now also works in a couple of corner cases, such as absence of ECG channel + absence of magnetometers in the final picks.

**Comment by jona-sassenhagen** [comment]
That's pretty cool.

File: mne/preprocessing/ica.py [hunk 6]
```diff
@@ -1436,6 +1441,12 @@ def plot_scores(self, scores, exclude=None, axhline=None,
         exclude : array_like of int
             The components marked for exclusion. If None (default), ICA.exclude
             will be used.
+        labels : str | list | 'ecg' | 'eog' | None
+            The labels to consider for the axes tests. Defaults to None.
+            If list, should match the outer shape of `scores`.
+            If 'ecg' or 'eog', the labels_ attributes will be looked up.
+            Note that '/' is used internally for sublabels specifying ECG and
```
Comment (jona-sassenhagen) [inline mne/preprocessing/ica.py]: Not "|"?

Also, should the docs to CORRMAP also be amended in this way?

Comment (dengemann) [inline mne/preprocessing/ica.py]: We never used this one and I find '/' more nice to read.

Comment (dengemann) [inline mne/preprocessing/ica.py]: What was the other question about corrmap?  Add docs there?

**Comment by dengemann** [comment]
Apparently we have coverage poroblems to overcome before merging. Will look into it later.

**Comment by dengemann** [comment]
rebased, fixed coverage issues.

**Comment by dengemann** [comment]
updated whatsnew

**Comment by dengemann** [comment]
ping can we merge? @agramfort @Eric89GXL ...

**Comment by jona-sassenhagen** [comment]
Cool stuff.

**Comment by Eric89GXL** [comment]
LGTM

**Comment by dengemann** [comment]
Thanks for your feedback. Happy ICA'ing.

**Pull request merged**